examples/custom_events.rs

```diff
@@ -18,30 +20,35 @@ fn main() -> Result<(), impl std::error::Error> {
     }
 
     SimpleLogger::new().init().unwrap();
-    let event_loop = EventLoopBuilder::<CustomEvent>::with_user_event()
-        .build()
-        .unwrap();
+    let event_loop = EventLoopBuilder::new().build().unwrap();
 
     let window = WindowBuilder::new()
         .with_title("A fantastic window!")
         .build(&event_loop)
         .unwrap();
 
-    // `EventLoopProxy` allows you to dispatch custom events to the main Winit event
-    // loop from any thread.
-    let event_loop_proxy = event_loop.create_proxy();
+    let waker = event_loop.create_proxy().waker();
+
+    let (sender, receiver) = mpsc::channel();
 
     std::thread::spawn(move || {
-        // Wake up the `event_loop` once every second and dispatch a custom event
-        // from a different thread.
+        // Dispatch a custom event from a different thread once every second,
+        // and wake the `event_loop` so that it can handle it.
         loop {
             std::thread::sleep(std::time::Duration::from_secs(1));
-            event_loop_proxy.send_event(CustomEvent::Timer).ok();
+            if sender.send(CustomEvent::Timer).is_err() {
+                break;
+            }
+            waker.wake_by_ref();
         }
     });
 
     event_loop.run(move |event, elwt| match event {
-        Event::UserEvent(event) => println!("user event: {event:?}"),
+        Event::NewEvents(_) => {
+            for event in receiver.iter() {
```

Comment:
blocking inside `NewEvents` feels not great, I would suggest `try_iter` here

Reply:
I'd prefer to have separate event for all of that tbf, so you kind of know that you woke up due to your event.

Reply:
> I would suggest `try_iter` here

I've changed, thanks!

> I'd prefer to have separate event for all of that tbf, so you kind of know that you woke up due to your event.

I think a `UserWoken` event is a bad idea, for that to be useful we'd have to dispatch it multiple times, once per invocation of `wake`.

This would mean that a valid chain of events is:
```
NewEvents(WaitCancelled { .. })
UserWoken
UserWoken
UserWoken
AboutToWait
```

Which seems counter-intuitive.

I'd be more inclined to adding it as something like `StartCause::UserWoken`, but that has its own set of problems.

Reply:
No, you invoke it only once, so you don't always poll your events if you don't have anything. Doing in `init` forces to do the checking every time, doing in separate event is the better way to put it.

The wakeup must be single per event loop. The WakeUp itself must be squashed. Basically how it's done in winit-next.

The point of being useful is that you could have complex sources of events and some may require locking to poll, so having a dedicated event sounds better, it also separates the logic and makes it more clear.

Reply:
> Doing in `init` forces to do the checking every time

Fair point.

My primary fear is that I don't want people to do:
```rust
Event::UserWoken => {
    let event = receiver.try_recv().unwrap();
    // Handle event
}
```

As they may then miss events, and I think that is more easily avoided if the wakeup is bundled into `NewEvents`.

Reply:
I don't understand what events are they going to miss. The order is `NewEvents`, and then `Wakeup` event, they are explicitly aware of such thing. The thing is that it also doesn't really matter, since they can poll in `NewEvents` and `Wakeup` event, because `NewEvents` and `Wakeup` are coming together, it's just a way to optimize/clean things a bit, given that winit already handled the part that you have new event.

Reply:
I don't want three calls to `waker.wake_by_ref()` to be guaranteed to result in three `Wakeup`/`UserWoken` events, but I fear the user might _assume_ that's how it works, and design their code around that assumption.

One solution to this would be to fold the event into `NewEvents`, as I'm proposing. Another is better documentation.

> it also doesn't really matter

No this is definitely nitpicking.